platform-sdk/docs/proposals/TSS-Block-Signing/tss-block-signing.md

```diff
@@ -0,0 +1,317 @@
+
+If we wanted a guarantee of generating a ledger signature on every block, we would need to add state that
+captures the accumulation of share signatures after consensus. If we can tolerate the possibility of failing to
+construct a ledger signature, we can avoid the introduction of state and process the share signature transactions
+pre-consensus. This allows us to construct signatures faster during healthy execution.
+
+### Design Decisions
+
+No Guaranteed Ledger Signature:
+
+- `TssShareSignatureTransaction` are handled pre-consensus in the pre-handle phase of transaction handling
+  - this is to allow the creation of signatures as fast as possible.
+- The `TssShareSignatureTransaction` are not stored in the state.
+
+This decision (coming from the Block Stream design) is to have faster signatures handled in pre-consensus and
+cover any missed signatures on blocks with Merkle Proofs using signatures from later blocks.
+
+#### Alternatives Considered
+
+Guaranteed Ledger Signature Generation:
+
+- requires storing share signatures in the state.
+- requires processing share signatures after consensus, introducing a delay in block signing.
+
+This was not selected for the following reasons:
+
+1. block proofs do not need to be identical between nodes.
+2. the probability of failure to sign a block is low.
+3. speed in signing a block matters since blocks are not published without signatures once they are signable.
+
+## Changes
+
+The changes are presented in the following order:
+
+1. Public API
+2. Component Architecture
+3. Core Behaviors
+4. Configuration
+5. Metrics
+6. Performance
+
+### Public API
+
+#### Services
+
+The `TssBaseService` is extended with the following API:
+
+```Java
+/**
+ * The TssBaseService will attempt to generate TSS key material for any set candidate roster, giving it a ledger id and
+ * the ability to generate ledger signatures that can be verified by the ledger id.  Once the candidate roster has
+ * received its full TSS key material, it can be made available for adoption by the platform.
+ * </p>
+ * The TssBaseService will also attempt to generate ledger signatures by aggregating share signatures produced by
+ * calling {@link #requestLedgerSignature(byte[])}.
+ *
+ */
+public interface TssBaseService {
+
+    ...
+
+    /**
+     * Requests a ledger signature on a message hash.  The ledger signature is computed asynchronously and returned
+     * to all consumers that have been registered through {@link #registerLedgerSignatureConsumer(Consumer)}.
+     *
+     * @param messageHash The hash of the message to be signed by the ledger.
+     */
+    void requestLedgerSignature(@Nonnull byte[] messageHash);
+
+    /**
+     * Registers a consumer of the message hash and the ledger signature on the message hash.
+     *
+     * @param consumer the consumer of ledger signatures and message hashes.
+     */
+    void registerLedgerSignatureConsumer(@Nonnull final BiConsumer<byte[], PairingSignature> consumer);
```

Comment:
Do we actually need a new type `PairingSignature`? Why not just `byte[]` or `Bytes`?

Reply:
It will not be a PBJ object unless someone else converts it to one because they need it in that form.    @mxtartaglia-sl , @AlfredoG87  What form should the ledger signature be communicated in?   byte[]?  This is communication to the block stream code for adding it to the to block.  I'm assuming that when we use the block signature, we'd be using the TSS-Library and have to convert it back to a PairingSignature in order to verify it with the library?

Reply:
I agree with @tinker-michaelj that an byte[] should be good.

Reply:
>@mxtartaglia-sl , @AlfredoG87 What form should the ledger signature be communicated in? byte[]?

@edward-swirldslabs, what are the expectations for the consumer? Should it write the signature to the block and be done with it, or should it do something else with it?

It is still unclear to me how external parties will read the signatures from the block. The block node is covered with a byte array if it uses the TSS library.

I prefer byte[] for all the reasons I explained in the meeting (an opaque byte array provides flexibility and allows curve adjustment or group assignment changes with less impact). Still, I am unsure if that would bring extra difficulties for external block consumers.

But if we use `PairingSignature`, it should be a protobuf type. Given that If the intention for the consumer is only to write the signature, then we are unnecessarily coupling the library to the consumer where they won't need to use it other than for calling the #toByteArray() method on the signature.

Reply:
Per our discussion yesterday, the ledger signature and the ledger id will be `byte[]` in the form that best fits consensus node and `TSS-Library` needs.   We'll publish a byte format for parsing the byte array to extract components by interested parties.   But for the block stream and consensus node, it's just a block of bytes.